thunder/benchmarks/benchmark_peft.py

```diff
@@ -0,0 +1,724 @@
+
+    # Set memory efficient settings
+    torch.backends.cuda.matmul.allow_tf32 = True
+    torch.backends.cudnn.allow_tf32 = True
+    torch.backends.cudnn.benchmark = True
+
+    # Calculate global batch size
+    gbs = args.mbs * args.grad_acc_steps * WORLD_SIZE
+    logger.info(
+        f"Global batch size: {gbs} (mbs: {args.mbs}, grad_acc_steps: {args.grad_acc_steps}, world_size: {WORLD_SIZE})"
+    )
+
+    # Initialize batch tracking
+    batches_processed = 0
+    total_tokens_processed = 0
+
+    seq_lengths = None
+    if args.var_seq_length:
+        random.seed(42)
+        seq_lengths = [random.randint(10, args.seq_length) for _ in range(gbs * args.max_steps)]
+        if args.verbose:
+            dist_print("variable-seq-lengths= " + str(seq_lengths))
+
+    start_ts = time.time()
+
+    # Initialize metrics tracking
+    iteration_times = []
+    max_allocated_memory = 0
+    total_tokens = 0
+    t0 = None  # For tracking throughput after warmup
+
+    logger.info("Loading tokenizer")
+    tokenizer = get_tokenizer(args.model, args.trust_remote_code)
+
+    # Load base model on meta device
+    logger.info(f"Loading base model on meta device...")
+    model = load_model(
+        args.model,
+        attn_implementation=args.attn_implementation,
+        fixed_num_hidden_layers=args.fixed_num_hidden_layers,
+        trust_remote_code=args.trust_remote_code,
+        verbose=args.verbose,
+    )
+    logger.info(f"Base model loaded on meta device")
+
+    # Configure model for static shapes
+    if hasattr(model, "config"):
+        model.config.use_cache = True
+        model.config.max_position_embeddings = args.seq_length
+        logger.info(f"Configured model for static shapes with sequence length: {args.seq_length}")
+
+    # Materialize the model on CUDA
+    model = model.to_empty(device=f"cuda:{LOCAL_RANK}")
+    model.apply(lambda m: m.reset_parameters() if hasattr(m, "reset_parameters") else None)
+
+    # Handle gradient checkpointing based on user preference
+    if hasattr(model, "gradient_checkpointing_enable"):
+        if args.gradient_checkpointing:
+            model.gradient_checkpointing_enable()
+            logger.info("Gradient checkpointing enabled")
+        else:
+            model.gradient_checkpointing_disable()
+            logger.info("Gradient checkpointing disabled")
+    else:
+        logger.info("Model does not support gradient checkpointing")
+
+    # Apply LoRA
+    logger.info(f"Applying LoRA to model")
+    model = setup_lora(model, verbose=args.verbose)
+    logger.info(f"LoRA applied to model")
+
+    # Ensure model is in training mode and verify gradients
+    model.train()
+    logger.info("Verifying gradient setup...")
+    for name, param in model.named_parameters():
+        if "lora" in name.lower():
+            if not param.requires_grad:
+                if args.verbose:
+                    logger.warning(f"LoRA parameter {name} does not require grad!")
```

Comment:
Should these be asserts instead? I think if the requires_grad was setup wrong, we shouldn't proceed with getting the numbers.